File: internal/runner/frames.go
package runner

import (
	"github.com/qualitymd/quality.md/internal/evaluation"
	"github.com/qualitymd/quality.md/internal/model"
)

// Deterministic frame generation. Frames are structural derivations of the
// model snapshot and planned scope, so the runner produces them itself; only
// judgment work goes to evaluators.

const (
	factorSynthesisGuidance = "protocol:factor-synthesis-default-v0"
	areaSynthesisGuidance   = "protocol:area-synthesis-default-v0"
)

func framePayload(kind evaluation.DataKind, body map[string]any) map[string]any {
	payload := map[string]any{
		"schemaVersion": evaluation.SchemaVersion,
		"kind":          string(kind),
	}
	for key, value := range body {
		payload[key] = value
	}
	return payload
}

func ratingLevelRefs(spec *model.Spec) []any {
	out := make([]any, 0, len(spec.RatingScale))
	for _, level := range spec.RatingScale {
		out = append(out, evaluation.RatingReference(level.Level))
	}
	return out
}

func evaluationFramePayload(spec *model.Spec, manifest Manifest) map[string]any {
	return framePayload(evaluation.DataKindEvaluationFrame, map[string]any{
		"subject": map[string]any{"modelLocator": manifest.Model},
		"inputs":  map[string]any{"ratingLevelIds": ratingLevelRefs(spec)},
		"derivedContext": map[string]any{
			"rigor":              "standard",
			"evaluationPolicies": []any{"source-as-data", "secret-redaction"},
		},
	})
}

func areaEvaluationFramePayload(area *PlannedArea) map[string]any {
	inputs := map[string]any{
		"localRequirementIds": anyStrings(area.LocalRequirements),
		"rootFactorIds":       anyStrings(area.RootFactors),
		"childAreaIds":        anyStrings(area.ChildAreas),
	}
	if area.Source != "" {
		inputs["sourceRefs"] = []any{area.Source}
	}
	return framePayload(evaluation.DataKindAreaEvaluationFrame, map[string]any{
		"subject": map[string]any{"areaId": area.Ref},
		"inputs":  inputs,
	})
}

func requirementEvaluationFramePayload(spec *model.Spec, req *PlannedRequirement) map[string]any {
	subject := map[string]any{"requirementId": req.Ref}
	if len(req.Factors) > 0 {
		subject["factorIds"] = anyStrings(req.Factors)
	}
	inputs := map[string]any{"ratingLevelIds": ratingLevelRefs(spec)}
	if req.Assessment != "" {
		inputs["requirementAssessmentBasis"] = req.Assessment
	}
	criteria := make([]any, 0, len(spec.RatingScale))
	for _, level := range spec.RatingScale {
		criterion := level.Criterion
		source := "model_default"
		if override, ok := req.Ratings[level.Level]; ok && override != "" {
			criterion = override
			source = "requirement_override"
		}
		if criterion == "" {
			continue
		}
		criteria = append(criteria, map[string]any{
			"ratingLevelId": evaluation.RatingReference(level.Level),
			"criterion":     criterion,
			"source":        source,
		})
	}
	body := map[string]any{"subject": subject, "inputs": inputs}
	if len(criteria) > 0 {
		body["derivedContext"] = map[string]any{"appliedRatingCriteria": criteria}
	}
	return framePayload(evaluation.DataKindRequirementEvaluationFrame, body)
}

func factorAnalysisFramePayload(factor *PlannedFactor) map[string]any {
	directRefs := make([]any, 0, len(factor.Requirements))
	for _, reqRef := range factor.Requirements {
		directRefs = append(directRefs, routineRefPayload("RequirementRatingResult", map[string]any{"requirementId": reqRef}, ""))
	}
	childRefs := make([]any, 0, len(factor.ChildFactors))
	for _, childRef := range factor.ChildFactors {
		childRefs = append(childRefs, routineRefPayload("FactorAnalysisResult", map[string]any{"factorId": childRef}, "localAndDescendantAnalysis"))
	}
	return framePayload(evaluation.DataKindFactorAnalysisFrame, map[string]any{
		"subject": map[string]any{"areaId": factor.Area.Reference(), "factorId": factor.Ref},
		"inputs": map[string]any{
			"directRequirementRatingRefs": directRefs,
			"childFactorAnalysisRefs":     childRefs,
		},
		"derivedContext": map[string]any{
			"synthesisGuidanceRef": factorSynthesisGuidance,
			"emptySignalPolicy":    "ignore_empty",
		},
	})
}

func areaAnalysisFramePayload(area *PlannedArea) map[string]any {
	factorRefs := make([]any, 0, len(area.RootFactors))
	for _, factorRef := range area.RootFactors {
		factorRefs = append(factorRefs, routineRefPayload("FactorAnalysisResult", map[string]any{"factorId": factorRef}, "localAndDescendantAnalysis"))
	}
	childRefs := make([]any, 0, len(area.ChildAreas))
	for _, childRef := range area.ChildAreas {
		childRefs = append(childRefs, routineRefPayload("AreaAnalysisResult", map[string]any{"areaId": childRef}, "localAndDescendantAnalysis"))
	}
	return framePayload(evaluation.DataKindAreaAnalysisFrame, map[string]any{
		"subject": map[string]any{"areaId": area.Ref},
		"inputs": map[string]any{
			"factorAnalysisRefs":    factorRefs,
			"childAreaAnalysisRefs": childRefs,
		},
		"derivedContext": map[string]any{
			"synthesisGuidanceRef": areaSynthesisGuidance,
			"emptySignalPolicy":    "ignore_empty",
		},
	})
}

func routineRefPayload(kind string, subject map[string]any, selector string) map[string]any {
	ref := map[string]any{"kind": kind, "subject": subject}
	if selector != "" {
		ref["selector"] = selector
	}
	return ref
}

File: src/domain/evaluation/frames.ts
import type { QualityModel } from "../model/model.ts"
import type { WorkUnit } from "./graph.ts"
import type { EvaluationPlan, PlannedFactor } from "./plan.ts"
import type { StoredPayload } from "./protocol.ts"

type JsonObject = Record<string, unknown>

const ratingIds = (model: QualityModel) => model.ratingScale.map((level) => `rating:${level.level}`)

const routineRef = (kind: string, subject: JsonObject, selector = "") => ({
  kind,
  ...(selector === "" ? {} : { selector }),
  subject,
})

const factorChildren = (factor: PlannedFactor, plan: EvaluationPlan) =>
  plan.factors
    .filter(
      (candidate) =>
        candidate.areaId === factor.areaId &&
        candidate.path.length === factor.path.length + 1 &&
        candidate.path.slice(0, -1).join("/") === factor.path.join("/"),
    )
    .map((candidate) => candidate.ref)

const factorRequirements = (factor: PlannedFactor, plan: EvaluationPlan) =>
  plan.requirements
    .filter((requirement) => requirement.factorIds.includes(factor.ref))
    .map((requirement) => requirement.ref)

const evaluationFrame = (model: QualityModel, modelPath: string): JsonObject => ({
  derivedContext: { evaluationPolicies: ["source-as-data", "secret-redaction"], rigor: "standard" },
  inputs: { ratingLevelIds: ratingIds(model) },
  kind: "EvaluationFrame",
  schemaVersion: 3,
  subject: { modelLocator: modelPath },
})

const areaEvaluationFrame = (area: EvaluationPlan["areas"][number]): JsonObject => ({
  inputs: {
    childAreaIds: area.childAreaIds,
    localRequirementIds: area.localRequirementIds,
    rootFactorIds: area.rootFactorIds,
    sourceRefs: [area.source],
  },
  kind: "AreaEvaluationFrame",
  schemaVersion: 3,
  subject: { areaId: area.ref },
})

const requirementEvaluationFrame = (
  model: QualityModel,
  requirement: EvaluationPlan["requirements"][number],
): JsonObject => ({
  derivedContext: {
    appliedRatingCriteria: model.ratingScale.map((level) => ({
      criterion: requirement.value.ratings?.[level.level] ?? level.criterion,
      ratingLevelId: `rating:${level.level}`,
      source:
        requirement.value.ratings?.[level.level] === undefined
          ? "model_default"
          : "requirement_override",
    })),
  },
  inputs: {
    ratingLevelIds: ratingIds(model),
    requirementAssessmentBasis: requirement.value.assessment,
  },
  kind: "RequirementEvaluationFrame",
  schemaVersion: 3,
  subject: { factorIds: requirement.factorIds, requirementId: requirement.ref },
})

export const initialFramePayloads = (
  model: QualityModel,
  plan: EvaluationPlan,
  modelPath: string,
): ReadonlyArray<StoredPayload> => [
  { workUnit: "frameEvaluation", payload: evaluationFrame(model, modelPath) },
  ...plan.areas.map((area) => ({
    workUnit: `frameAreaEvaluation:${area.ref}`,
    payload: areaEvaluationFrame(area),
  })),
  ...plan.requirements.map((requirement) => ({
    workUnit: `frameRequirementEvaluation:${requirement.ref}`,
    payload: requirementEvaluationFrame(model, requirement),
  })),
]

export const deterministicPayload = (
  unit: WorkUnit,
  model: QualityModel,
  plan: EvaluationPlan,
  modelPath: string,
): JsonObject => {
  if (unit.kind === "frameEvaluation") return evaluationFrame(model, modelPath)
  if (unit.kind === "frameAreaEvaluation") {
    const area = plan.areas.find((entry) => entry.ref === unit.subject)!
    return {
      inputs: {
        childAreaIds: area.childAreaIds,
        localRequirementIds: area.localRequirementIds,
        rootFactorIds: area.rootFactorIds,
        ...(area.source === "" ? {} : { sourceRefs: [area.source] }),
      },
      kind: "AreaEvaluationFrame",
      schemaVersion: 3,
      subject: { areaId: area.ref },
    }
  }
  if (unit.kind === "frameRequirementEvaluation") {
    const requirement = plan.requirements.find((entry) => entry.ref === unit.subject)!
    const criteria = model.ratingScale.flatMap((level) => {
      const criterion = requirement.value.ratings?.[level.level] ?? level.criterion
      if (criterion === "") return []
      return [
        {
          criterion,
          ratingLevelId: `rating:${level.level}`,
          source:
            requirement.value.ratings?.[level.level] === undefined
              ? "model_default"
              : "requirement_override",
        },
      ]
    })
    return {
      ...(criteria.length === 0 ? {} : { derivedContext: { appliedRatingCriteria: criteria } }),
      inputs: {
        ratingLevelIds: ratingIds(model),
        ...(requirement.value.assessment === ""
          ? {}
          : { requirementAssessmentBasis: requirement.value.assessment }),
      },
      kind: "RequirementEvaluationFrame",
      schemaVersion: 3,
      subject: {
        ...(requirement.factorIds.length === 0 ? {} : { factorIds: requirement.factorIds }),
        requirementId: requirement.ref,
      },
    }
  }
  if (unit.kind === "frameFactorAnalysis") {
    const factor = plan.factors.find((entry) => entry.ref === unit.subject)!
    return {
      derivedContext: {
        emptySignalPolicy: "ignore_empty",
        synthesisGuidanceRef: "protocol:factor-synthesis-default-v0",
      },
      inputs: {
        childFactorAnalysisRefs: factorChildren(factor, plan).map((ref) =>
          routineRef("FactorAnalysisResult", { factorId: ref }, "localAndDescendantAnalysis"),
        ),
        directRequirementRatingRefs: factorRequirements(factor, plan).map((ref) =>
          routineRef("RequirementRatingResult", { requirementId: ref }),
        ),
      },
      kind: "FactorAnalysisFrame",
      schemaVersion: 3,
      subject: { areaId: factor.areaId, factorId: factor.ref },
    }
  }
  if (unit.kind === "frameAreaAnalysis") {
    const area = plan.areas.find((entry) => entry.ref === unit.subject)!
    return {
      derivedContext: {
        emptySignalPolicy: "ignore_empty",
        synthesisGuidanceRef: "protocol:area-synthesis-default-v0",
      },
      inputs: {
        childAreaAnalysisRefs: area.childAreaIds.map((ref) =>
          routineRef("AreaAnalysisResult", { areaId: ref }, "localAndDescendantAnalysis"),
        ),
        factorAnalysisRefs: area.rootFactorIds.map((ref) =>
          routineRef("FactorAnalysisResult", { factorId: ref }, "localAndDescendantAnalysis"),
        ),
      },
      kind: "AreaAnalysisFrame",
      schemaVersion: 3,
      subject: { areaId: area.ref },
    }
  }
  throw new Error(`no deterministic payload for ${unit.id}`)
}
